### PR TITLE
[For review only] 12327 Pdf viewers

### PR DESCRIPTION
**Pull request by kriswest**
**Resolves issue [12327](https://chartiq.kanbanize.com/ctrl_board/21/cards/12327/details)**
For review only. Two PDF viewers based on PDF.js and viewer.js, which FInsemble customisation (component state, linker and drag and drop sharing, PDF-aware native preloads).

**Description of change**
- pdfJS component: PDF.js based PDF viewer [README](https://github.com/ChartIQ/finsemble-seed/blob/PDFViewers/src/components/pdfJs/README.md)
- viewerJS component: Viewer.js based PDF viewer [README](https://github.com/ChartIQ/finsemble-seed/blob/PDFViewers/src/components/viewerJs/README.md)
- pdfTester component: Test component for reviewing the two PDF viewers [README](https://github.com/ChartIQ/finsemble-seed/tree/PDFViewers/src/components/pdfTester)

**Description of testing**
-Launch pdfTester and play on!

### CODE REVIEW

File: build/webpack/webpack.components.entries.json [hunk 1]
```diff
@@ -1 +1,14 @@
-{}
\ No newline at end of file
+{
```
Comment (NinjaManatee) [inline build/webpack/webpack.components.entries.json]: Use finsemble.webpack.json and put it in in the component folder. This makes things more modular.

File: build/webpack/webpack.preloads.entries.json [hunk 2]
```diff
@@ -6,5 +6,9 @@
 	"zoom": {
 		"output": "preloads/zoom",
 		"entry": "./src-built-in/preloads/zoom.js"
+	},
```
Comment (NinjaManatee) [inline build/webpack/webpack.preloads.entries.json]: I have a PR to make finsemble.webpack.json work for preloads too (https://github.com/ChartIQ/finsemble-seed/pull/333).

File: src/components/pdfJs/README.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+```
+{
+	...
+	"pdfJsNativeOverrides": {
+		"output": "preloads/pdfJsNativeOverrides",
+		"entry": "./src/components/pdfJs/pdfJsNativeOverrides.js"
+	}
+}
+```
+4. Add the following configuration to your _/configs/application/components.json_ file:
```
Comment (NinjaManatee) [inline src/components/pdfJs/README.md]: Move this config to a _src/components/pdfJs/config.json_.

File: src/components/pdfJs/README.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
+## PDF.js PDF Viewer Component ##
+This component adds support for viewing PDF files inside a Finsemble PDF viewer, based on [PDF.js](https://github.com/mozilla/pdf.js).
+
+![](./screenshot.png)
+
+The viewer can be launched for a specified PDF, preserves its state in workspaces and supports context linking on color channels or drag and drop, via topic names `url` and `pdf`. The payload can either be string or an object with key `url`, indicating which PDF URL to load. 
+
+### Installation ###
+1. Add the `pdfJS` component folder to your _/src/components_ directory.
+2. Add the component to your webpack build at _/build/webpack/webpack.components.entries.json_:
```
Comment (NinjaManatee) [inline src/components/pdfJs/README.md]: With the changes above, steps 2 and 3 become unnecessary.

File: src/components/pdfTester/README.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+Please review [pdfTester.html](./pdfTester.html) and [pdfTester.js](./pdfTester.js) for other examples of spawning or context sharing with the viewers.
+
+### Installation ###
+1. Add the `pdfTester` component folder to your _/src/components_ directory.
+2. Add the component to your webpack build at _/build/webpack/webpack.components.entries.json_: 
+```
+{
+    ...
+    "pdfTester": {
```
Comment (NinjaManatee) [inline src/components/pdfTester/README.md]: Move to finsemble.webpack.json in _src/components/pdfTester_

File: src/components/pdfTester/README.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        "entry": "./src/components/pdfTester/pdfTester.js"
+    },
+    ...
+}
+```
+3. Install either or both of the [PDF.js](../pdfJs/README.md) or [Viewer.js](../viewerJs/README.md) PDF viewer components (as per their own instructions).
+4. Add the following configuration to your _/configs/application/components.json_ file:
+```
+{
+    "components": {
```
Comment (NinjaManatee) [inline src/components/pdfTester/README.md]: move to _src/compnents/pdfTester/config.json_

File: src/components/viewerJs/viewerJs.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+					else { FSBL.Clients.Logger.error("Error on drag and drop (url): ", err); }
+				}
+			}
+		]
+	});
+};
+
+const init = function (){
+	//Check for existing component state
+	let state =
```
Comment (WatsonCIQ) [inline src/components/viewerJs/viewerJs.js]: Seems that **state** is set but then doesn't seem to be read anywhere, is it needed?

File: src/components/viewerJs/viewerJsNativeOverrides.js [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+	if (parsedUrl.pathname.toLowerCase().endsWith(".pdf")) {
+		//Handle a PDF
+		return handlePdfUrl(urlToOpen, name, specs, replace);
+	} else {
+		//Handle other URL
+		return handleOtherUrl(urlToOpen, name, specs, replace);
+	}
+}
+
+const handlePdfUrl = function(urlToOpen, name, specs, replace) {
```
Comment (WatsonCIQ) [inline src/components/viewerJs/viewerJsNativeOverrides.js]: specs and replace params are unused / never read, are they needed?

Comment (kriswest) [reply src/components/viewerJs/viewerJsNativeOverrides.js]: Left in as thats the signature of the built-in in JS function we're overloading

File: src/components/viewerJs/viewerJs.js (hunk 7 above)
Comment (kriswest) [reply src/components/viewerJs/viewerJs.js]: Yep can be removed

**Comment by kriswest** [comment]
Hey @Mark Valeiras <mark@chartiq.com>, one thing that may need doing to
this is having PDF.js included via an NPM module (probably
https://www.npmjs.com/package/pdfjs-dist)  instead of included in the
component. Its not criitical if its moved to its own repo but might make it
easier to maintian long term. Essentially the build script would copy the
viewer from node_modules directory, add our code and build as normal...

I started doing this, realised it was going to be a big pain until it was
moved to an fpe repo, started looking at that and then realised the webpack
scripts in the FPE repos looked derived from our (somewhat complex)
examples and then moved on (leaving myself a note to send you this message
:-) ).

I'd go ahead and move it as is before attempting the above. I'd also
consider giving it a super simple webpack config - not my forte, but Chris
Watson might be up for taking a quick look once its moved over and working
in current form. I've had a quick word with him to confirm that (and have
mentioned much much simpler webpack scripts - which I think you two should
talk about anyway).

K

**Pull request closed without merging**